### ISSUE TITLE
Turn off lights when display goes to sleep on macOS

### ISSUE THREAD

**Issue by danielcompton** [psieg/Lightpack#98]
I find that the Lightpack lights stay on when my display or computer goes to sleep. It would be great if they turned off along with my screen.

**Comment by psieg** [comment]
You can't by any chance point to documentation that shows how to get such events using macOS API?

**Comment by psieg** [comment]
ref #82

**Comment by danielcompton** [comment]
Here's a code sample a friend used:

```
NSWorkspace.shared().notificationCenter.addObserver(self, selector: Selector("sleepListener"), name: NSNotification.Name.NSWorkspaceWillSleep, object: nil)
        
        
          NSWorkspace.shared().notificationCenter.addObserver(self, selector: Selector("powerOffListener"), name: NSNotification.Name.NSWorkspaceWillPowerOff, object: nil)
        NSWorkspace.shared().notificationCenter.addObserver(self, selector: Selector("wakeUpListener"), name: NSNotification.Name.NSWorkspaceDidWake, object: nil)
        NSWorkspace.shared().notificationCenter.addObserver(self, selector: Selector("displaySleepListener"), name: NSNotification.Name.NSWorkspaceScreensDidSleep, object: nil)
```

The selector looks like: `@objc func sleepListener() {...`

In general, what you need to do is get notified from [NSWorkSpace](https://developer.apple.com/reference/appkit/nsworkspace) events about sleep and wake. These look relevant, although some are probably unnecessary if you have others:

* `NSWorkspaceDidWake`
* `NSWorkspaceWillPowerOff`
* `NSWorkspaceWillSleep`
* `NSWorkspaceScreensDidSleep`
* `NSWorkspaceScreensDidWake`

**Comment by psieg** [comment]
The corresponding class is the SessionChangeDetector, that needs a mac implementation. I have no mac available and development using Travis CI runs is too complicated. Anyone with OS X development experience is welcome to help

**Comment by psieg** [comment]
@danielcompton Do you want to try this for the next release? It seems like an easy thing, given all you need to do is adding the listeners and emit the `sessionChangeDetected` signal from [`SessionChangeDetector`](https://github.com/psieg/Lightpack/blob/master/Software/src/SessionChangeDetector.hpp)

**Comment by danielcompton** [comment]
I'd definitely be interested to work on this, but don't hold up a release for me. I could see doing this in Objective-C, but translating across into C++ and QT is a bit more foreign for me.

**Comment by psieg** [comment]
:+1:
You can skip the QT part if you like, that's rather straightforward (for someone who knows it already). The Mac specific things are what I can't do

**Comment by psieg** [comment]
Should be fixed since #225
